### PR TITLE
QAOA parameter setting analysis using IBM Quantum data

### PR DESCRIPTION
**Pull request by anurag-r20**
## Summary
This PR adds adds a QAOA parameter setting analysis workflow using IBM Quantum data, 
## How to Run Experiments
- Run `examples/IBM_QAOA/notebooks/Analysis.ipynb`
- Analyze results and compare between different QAOA parameter setting strategies 
- Check generated plots for trends and performance

## Notes
- Notebook outputs are included for reference
- Happy to refactor this PR if needed

### CODE REVIEW

**Comment by anurag-r20** [comment]
@bernalde

**Comment by anurag-r20** [comment]
@copilot+gpt-5.3-codex Review this PR. Do not make any commits/change any code without asking first!

File: tests/test_stochastic_benchmark_pandas.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        'param1': [10, 20, 30, 40],
+        'train': [0, 1, 0, 1],
+        # RESTORED: Add dummy confidence interval columns required by virtual_best
+        names.param2filename({'Key': 'response', 'ConfInt': 'lower'}, ''): [0.4, 0.5, 0.6, 0.7],
+        names.param2filename({'Key': 'response', 'ConfInt': 'upper'}, ''): [0.6, 0.7, 0.8, 0.9]
+    })
+    sb.stat_params = StatsParameters(metrics=['response'], stats_measures=[Median()])
+    
+    # RESTORED: Global setup for 'here'
+    sb.here = type('obj', (object,), {'checkpoints': '/tmp'})
```
Comment (bernalde) [inline tests/test_stochastic_benchmark_pandas.py]: Blocking: This fixture constructs a `stochastic_benchmark` object but the tests pass it directly to `VirtualBestBaseline` and `ProjectionExperiment`. Those classes expect the experiment-parameter contract, including `checkpoint_path`, so `python run_tests.py unit` fails six tests with `AttributeError: 'stochastic_benchmark' object has no attribute 'checkpoint_path'`. Build an `experiments.ExperimentParameters` fixture, or call `sb.get_experiment_parameters()` after populating the required state, and use `tmp_path` for the checkpoint directory so these tests exercise the production object shape.

File: examples/IBM_QAOA/src/Processing.py [hunk 2]
```diff
@@ -0,0 +1,954 @@
+''' This file is for processing QAOA data files generated by IBM using QAOA.
```
Comment (bernalde) [inline examples/IBM_QAOA/src/Processing.py:1]: Blocking: This new IBM QAOA processing surface is not covered by active collected tests. The PR adds fixtures under `tests/fixtures/ibm_qaoa`, but the comprehensive processing tests are under `examples/IBM_QAOA/Archive/ibm_qaoa_processing_tests.py`, which pytest does not collect and which imports the old `ibm_qaoa_processing` module path. Please move focused tests into `tests/test_ibm_qaoa_processing.py` for the current modules, including `QAOAHardware.load_hardware_instance`, `QAOATraining.load_training_instance`, min/max cut lookup, approximation-ratio conversion, and training-cost resolution.

File: src/training.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ def br(df):
             smooth,
         )
 
-    vb = df.groupby(groupby).apply(br, include_groups=False).reset_index()
+    vb = df.groupby(groupby).apply(br).reset_index() # include_groups=False, Pandas Version Error
```
Comment (bernalde) [inline src/training.py]: Nonblocking: Removing `include_groups=False` brings back the pandas `DataFrameGroupBy.apply operated on the grouping columns` FutureWarning under the repo's required pandas 2.3.x, and pandas will change this behavior in a future release. Since `requirements.txt` already requires `pandas>=2.3`, please restore `include_groups=False` here and in the parallel `interpolate.py`/`stats.py` groupby calls, or add a small compatibility helper if older pandas support is still needed.

File: examples/IBM_QAOA/src/simulation_validation.py [hunk 4]
```diff
@@ -0,0 +1,3888 @@
...
+    best_prefix_metrics,
+    extract_minmax_args,
+    get_minmax,
+    load_maxcut_instance_context,
+    maxcut_approximation_ratio,
+    maxcut_energy_from_bitstring,
+)
+
+
+DEFAULT_MAIN_REPO = Path("/mnt/c/Users/rames102/Desktop/QAOA-Parameter-Setting")
```
Comment (bernalde) [inline examples/IBM_QAOA/src/simulation_validation.py]: Nonblocking: This default points to a contributor-specific desktop path, so the script/notebook workflow leaves the documented path for any other checkout unless the user already knows which env var or CLI flag to override. Prefer no personal absolute default: derive paths from repo-relative locations where possible, or make the argument required and document `QAOA_PARAMETER_SETTING_ROOT` and `QAOA_TRAINING_PIPELINE_ROOT` in the notebook README.

File: examples/IBM_QAOA/notebooks/README.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+- `Simulation_Method_Validation_and_WS.ipynb`
+  - Simulation-method validation and Window Sticker notebook for the FA/PT-style PSS campaign.
+  - Loads campaign frontier data from `../results/pss_window_sticker/<result_tag>`.
+  - Runs the stochastic-benchmark Window Sticker workflow through `../src/simulation_validation.py`.
+  - Uses plotting and reporting helpers from `../src/utils.py`; this notebook should remain orchestration/display only and should not define local helper functions.
+  - Keeps the proxy-vs-exact calibration as the power-law/through-origin diagnostic plot used by the current experiment.
+
+## Notes
+
+- The notebook expects local data directories/instance directories to be available (paths are currently set inside the notebook).
```
Comment (bernalde) [inline examples/IBM_QAOA/notebooks/README.md:26]: Question: The new notebooks/scripts require external projects and packages (`qaoa_training_pipeline`, `qiskit`, `qiskit-aer`, and the QAOA-Parameter-Setting data checkout), but the repo's `requirements-examples.txt` only installs `scikit-learn` and this README does not give setup commands. Should this workflow add an IBM_QAOA-specific requirements file or a setup section with the required env vars and dependency installation?

**Review by bernalde — CHANGES_REQUESTED** [review]
Blocking issues:\n- The documented unit suite currently fails in the new pandas stochastic-benchmark tests because the fixture passes a `stochastic_benchmark` instance where the experiment classes require the `ExperimentParameters` contract, including `checkpoint_path`.\n- The new IBM_QAOA processing modules add a large user-facing parsing/analysis surface without active collected tests for the current modules.\n\nNonblocking issues:\n- The pandas `groupby.apply` compatibility changes reintroduce FutureWarnings under the repo's required pandas 2.3.x and should be made future-safe.\n- The IBM_QAOA script defaults still point at contributor-local absolute paths, which makes the documented workflow hard to reproduce outside that machine.\n\nQuestions:\n- Should the IBM_QAOA workflow add an example-specific requirements/setup section for qiskit, qiskit-aer, qaoa_training_pipeline, and the QAOA-Parameter-Setting checkout?\n\nTests run and outcomes:\n- `python -m pytest tests/test_interpolate_pandas.py tests/test_stats_pandas.py tests/test_stochastic_benchmark_pandas.py tests/test_training_pandas.py tests/test_df_utils.py tests/test_stats.py -q` on the base Python 3.13: collection failed because the base env lacks `multiprocess`; reran in the documented CI conda env.\n- `conda run -n stochastic-benchmark-ci-py310 python -m pytest tests/test_interpolate_pandas.py tests/test_stats_pandas.py tests/test_stochastic_benchmark_pandas.py tests/test_training_pandas.py tests/test_df_utils.py tests/test_stats.py -q`: 37 passed, 6 failed, all in `tests/test_stochastic_benchmark_pandas.py`.\n- `conda run -n stochastic-benchmark-ci-py310 python run_tests.py unit`: 217 passed, 6 failed, same failures.\n- `conda run -n stochastic-benchmark-ci-py310 python run_tests.py integration`: 11 passed.\n- `conda run -n stochastic-benchmark-ci-py310 python -m py_compile examples/IBM_QAOA/run_prepare_pss_campaign.py examples/IBM_QAOA/src/Processing.py examples/IBM_QAOA/src/approx_ratio_calc.py examples/IBM_QAOA/src/simulation_validation.py examples/IBM_QAOA/src/utils.py`: passed.\n- `conda run -n stochastic-benchmark-ci-py310 python -c \"import sys; sys.path.insert(0, 'examples/IBM_QAOA'); import src.Processing, src.approx_ratio_calc, src.utils, src.simulation_validation; print('imports ok')\"`: passed.\n- `conda run -n stochastic-benchmark-ci-py310 python examples/IBM_QAOA/run_prepare_pss_campaign.py --help`: passed.\n\nThe PR should not be merged as-is. I would not merge this until the blocking issues above are addressed.

**Review by bernalde — CHANGES_REQUESTED** [review]
Blocking issues:
- The existing `tests/test_stochastic_benchmark_pandas.py` blocking thread still applies. The full documented test suite fails six newly added tests because they pass a `stochastic_benchmark` instance directly to `VirtualBestBaseline` and `ProjectionExperiment`, while those classes require the experiment-parameter contract with `checkpoint_path`.
- The existing IBM QAOA processing coverage blocking thread still applies. The PR adds fixtures and documents `tests/test_ibm_qaoa_processing.py`, but that file is absent; the only comprehensive processing tests are archived, uncollected, and currently fail during collection.
- The PR is currently merge-conflicting with `main` (`mergeable: CONFLICTING`), so it cannot be merged as-is.

Nonblocking issues:
- The existing `include_groups=False` thread still applies. The PR reintroduces pandas groupby warnings under the required pandas 2.3.x range and should either restore `include_groups=False` or add an explicit compatibility helper.
- The existing personal absolute path/defaults comment still applies for the IBM QAOA workflow.

Questions:
- The existing setup/dependency question for the IBM QAOA notebooks/scripts remains open.

Tests run and outcomes:
- `conda run -n stochastic-benchmark-ci-py310 python -m pytest tests/ -q`: failed, 6 failed and 228 passed. All failures are in `tests/test_stochastic_benchmark_pandas.py`.
- `conda run -n stochastic-benchmark-ci-py310 python -m pytest tests/test_interpolate_pandas.py tests/test_stats_pandas.py tests/test_stochastic_benchmark_pandas.py tests/test_training_pandas.py tests/test_df_utils.py tests/test_stats.py -q`: failed, 6 failed and 37 passed. Same `checkpoint_path` failures.
- `conda run -n stochastic-benchmark-ci-py310 python -m pytest tests/test_interpolate.py tests/test_training.py tests/test_stats.py -q`: passed, 45 passed.
- `conda run -n stochastic-benchmark-ci-py310 python -m pytest examples/IBM_QAOA/Archive/ibm_qaoa_processing_tests.py -q`: failed during collection with `NameError: name 'Dict' is not defined`, and this path is outside configured pytest collection.
- `conda run -n stochastic-benchmark-ci-py310 flake8 src --count --select=E9,F63,F7,F82 --show-source --statistics`: passed, 0 critical findings.
- `conda run -n stochastic-benchmark-ci-py310 flake8 src --count --exit-zero --max-complexity=10 --max-line-length=120 --statistics`: exited 0 with 142 reported style warnings, consistent with the command's warning-only behavior.
- `gh pr checks 50 --repo usra-riacs/stochastic-benchmark`: no checks reported on the PR branch.

I would not merge this until the blocking issues above are addressed.

**Comment by anurag-r20** [comment]
@bernalde All checks passed, Added documentation for dependencies

**Review by bernalde — CHANGES_REQUESTED** [review]
Re-review at `3bf76ec`. Thanks for the substantial rework — three of the four blockers from the earlier rounds are genuinely fixed, and the new `tests/test_ibm_qaoa_processing.py` is the right shape.

GitHub cannot render hunks for this PR (`gh pr diff` returns `HTTP 406: diff exceeded the maximum number of lines (20000)`, and the pull-files endpoint only returns `patch` for the first 42 files — none of the files below). All findings are therefore in this body with stable markers rather than inline.

**Status of the earlier threads at this head**

- `tests/test_stochastic_benchmark_pandas.py` / `checkpoint_path`: **resolved.** The fixture now feeds `get_experiment_parameters()`; all 7 tests pass.
- IBM QAOA processing coverage: **resolved.** `tests/test_ibm_qaoa_processing.py` exists and its 8 tests pass against the current `src/` modules.
- Merge conflict with `main`: **resolved.** `mergeable: true`, and `git merge-tree --write-tree origin/main 3bf76ec` produces a tree with no conflicts.
- Personal absolute path default in `simulation_validation.py`: **partly resolved.** The hardcoded `/mnt/c/Users/...` default is gone, but the replacement resolves to the wrong directory — see the first blocker.
- `include_groups=False`: **not resolved,** and I now have evidence it should not have been removed at all. Escalated to blocking below.
- External-dependency setup question: **partly answered.** See the question at the end.

## Blocking

### Default external-repo paths resolve inside this checkout, not beside it 

`examples/IBM_QAOA/src/simulation_validation.py:32-33`

```python
DEFAULT_MAIN_REPO = Path(__file__).resolve().parents[3] / "QAOA-Parameter-Setting"
DEFAULT_PIPELINE_REPO = Path(__file__).resolve().parents[3] / "qaoa_training_pipeline"
```

The file lives at `<repo>/examples/IBM_QAOA/src/`, so `parents[3]` is `<repo>` itself. The defaults therefore point at `<repo>/QAOA-Parameter-Setting` and `<repo>/qaoa_training_pipeline` — inside the `stochastic-benchmark` working tree.

Both new READMEs document the opposite layout:

```text
<workspace>/stochastic-benchmark
<workspace>/QAOA-Parameter-Setting
<workspace>/qaoa_training_pipeline
```

(`examples/IBM_QAOA/notebooks/README.md:43-50` "looks for sibling checkouts next to this repository", and `examples/IBM_QAOA/src/README.md:34,49-55`.) Anyone who follows the documented layout gets a `FileNotFoundError` on the default path and has to discover the env vars to recover, which is most of what the earlier thread asked to avoid. Use `parents[4]` for the two sibling repos; `DEFAULT_INSTANCE_CACHE_ROOT` on line 34 is correct as written (`parents[1]` → `examples/IBM_QAOA/data/generated_instances`, matching the README).

### The four new notebooks are not registered in `examples/tutorials.json` 

`examples/tutorials.json` is this repository's registry for example notebooks: `scripts/verify_tutorials.py` reads it, `tests/test_verify_tutorials.py` asserts on it, and the `tutorial-smoke` CI job runs from it. Every notebook under `examples/` on `main` is registered — 7 of 7, verified by diffing `git ls-tree -r --name-only main -- examples | grep '\.ipynb$'` against the manifest paths.

This PR adds four and registers none:

```
examples/IBM_QAOA/notebooks/Analysis.ipynb
examples/IBM_QAOA/notebooks/Simulation_Method_Validation_and_WS.ipynb
examples/IBM_QAOA/Archive/ibm_qaoa_analysis.ipynb
examples/IBM_QAOA/Archive/ibm_qaoa_analysis_hardware.ipynb
```

`verify_tutorials.py` only walks the manifest, so nothing fails — the notebooks are simply invisible to the tutorial contract. The `external` category exists for exactly this case; please add entries with `"category": "external"` and a `reason` naming the required `QAOA-Parameter-Setting` / `qaoa_training_pipeline` checkouts (or drop the `Archive/` notebooks entirely per the Archive finding below).

### `Analysis.ipynb` ships a kernel-crash output on its final plot cell 

`examples/IBM_QAOA/notebooks/Analysis.ipynb`

The PR description says "Run `examples/IBM_QAOA/notebooks/Analysis.ipynb`", "Notebook outputs are included for reference", and "Check generated plots for trends and performance". The committed outputs do not support that. Code cell index 28 — the recommendation plot, which is the notebook's headline result — carries an `error` output:

```
The Kernel crashed while executing code in the current cell or a previous cell.
```

That cell is `build_recommendation_data` → `plot_ibm_qaoa_recommendation` → `print_budget_recommendations`, and its `execution_count` is `null`, so it produced no figure and no budget table. The rest of the notebook is also not a single top-to-bottom run: `execution_count` goes `1`, then `26 … 46`, then `null, null` for the last two cells.

Please re-execute the notebook cleanly and commit outputs from that run, or if the crash is a resource limit on the recommendation cell, say so in the notebook and in the README and strip the error output rather than committing it. `Simulation_Method_Validation_and_WS.ipynb` has the same restart pattern (counts `275…281, 282, 289`, last cell unexecuted) though no error output; worth regenerating together.

### Nautilus manifests and runner default to this PR's own feature branch 

Five places pin `QAOA_Parameter_Setting_IBM`:

- `examples/IBM_QAOA/nautilus/run_simulation_validation.sh:11`
- `examples/IBM_QAOA/nautilus/dev-pod.yaml:42`
- `examples/IBM_QAOA/nautilus/simulation-validation-job.yaml:109`
- `examples/IBM_QAOA/nautilus/simulation-validation-sharded-job.yaml:72`
- `examples/IBM_QAOA/nautilus/simulation-validation-finalize-job.yaml:62`

Once this merges the branch is normally deleted, and `git clone --branch "${STOCHASTIC_BENCHMARK_BRANCH}"` under `set -euo pipefail` then fails at job startup — every manifest in the merged tree is broken on day one. Default `STOCHASTIC_BENCHMARK_BRANCH` to `main` in the script and in the four manifests (keeping the env-var override for branch testing), and update `examples/IBM_QAOA/nautilus/README.md:28`, which still instructs `git push upstream QAOA_Parameter_Setting_IBM`.

The token handling in these manifests looks right, for the record: `secretKeyRef` with `optional: true`, written to a `0600` credential file, never passed on a command line.

### `src/training.py` drops `include_groups=False`, regressing `main` 

`src/training.py:141,153`

```python
            ) # include_groups=False # Pandas Version Error
```

`main` currently passes `include_groups=False` in both `split_train_test` calls; this PR removes them. The "Pandas Version Error" justification does not hold, and I checked it directly rather than reasoning from the changelog:

- pandas 2.3.3 (the floor in `requirements.txt`, `pandas>=2.3`): `include_groups=False` is accepted and silent; omitting it emits `FutureWarning`; the two results are `.equals()`-identical.
- pandas 3.0.5: both forms are accepted and identical.

So removing it buys nothing and reintroduces the warning on the supported floor. It is also inconsistent within the same file — lines 111 and 328 still pass it, as do `interpolate.py`, `stochastic_benchmark.py`, and `cross_validation.py`, and `tests/test_df_utils.py:127-150` asserts the kwarg is forwarded. Restoring it on both call sites keeps the suite green (I applied it locally: 332 passed at this head, `tests/test_training.py` 24 passed with `FutureWarning` promoted to visible). Please restore both, and drop the comment.

### Four zero-byte test files are wired into `run_tests.py unit` 

```
tests/test_interpolate_groups.py            0 bytes
tests/test_stats_groups.py                  0 bytes
tests/test_stochastic_benchmark_groups.py   0 bytes
tests/test_training_groups.py               0 bytes
```

`run_tests.py` globs `tests/test_*.py`, so all four appear on the documented `python run_tests.py unit` command line and contribute zero tests. Whatever the "move Pandas group tests to main tests directory" commit intended, the group-variant content did not arrive. Either add the tests or delete the four files; an empty `test_*.py` advertises coverage that does not exist.

## Nonblocking

### `examples/IBM_QAOA/Archive/` commits ~10k lines of non-working dead code 

1.8 MB across nine files. It is not merely unused, it is broken and carries things that should not land:

- `flake8 --select=E9,F63,F7,F82` reports **23 F821 errors**, all in `Archive/ibm_qaoa_processing.py`: `Dict` (18) and `Any` (4) are never imported, and line 887 calls `validate_edge_map`, which does not exist anywhere in the file — the defined function is `validate_edge_map_program_qubits` at line 715, with a different signature. The rest of `examples/IBM_QAOA` is clean (0 critical findings).
- `Archive/transfer_hardware_data.py:32` and `Archive/transfer_simulation_data.py:224,226` hardcode `/mnt/c/Users/rames102/Desktop/QAOA-Parameter-Setting/...` — the same class of path the earlier review asked to remove.
- `Archive/test_approx_ratio_fix.py` is a `test_*.py` file; `pytest examples/IBM_QAOA/Archive` collects 3 tests from it. It escapes CI only because `testpaths = ["tests"]`.
- `Archive/ibm_qaoa_analysis_hardware.ipynb` commits a `NameError: name 'selected_p' is not defined` output.

Git history already preserves this code. I'd drop the directory; if it must stay, at minimum remove the contributor paths, fix or delete the broken module, and rename the stray test file so it cannot be collected.

### Committed notebook outputs still contain contributor absolute paths 

`examples/IBM_QAOA/notebooks/Simulation_Method_Validation_and_WS.ipynb` prints `/Users/rames102/Desktop/stochastic-benchmark/...` and `/Users/rames102/Desktop/QAOA-Parameter-Setting/data/hardware` in its outputs (also truncated into several DataFrame HTML tables). The code defaults are now clean, so this is just stale output; it will go away with the re-execution asked for above.

### `tests/test_stats.py` change is a removed trailing newline 

The whole diff for this file is `\ No newline at end of file` on the last line. Please restore the newline and drop the file from the PR.

### `best_recommended` behavior change is unrelated and untested 

`src/training.py:191` changes `vb.groupby(resource_col).mean()` to `.mean(numeric_only=True)`. `main`'s suite is green on pandas 3.0.5 without it (426 passed), so nothing in the repo demonstrates the problem it fixes, and no test covers the new behavior. It also changes the failure mode for a non-numeric parameter column from a `TypeError` at the aggregation to a `KeyError` at `br[parameter_names + additional_cols]`. Either split it into its own change with a test that fails before and passes after, or drop it from this PR.

### flake8 findings in the new example code 

`flake8 examples/IBM_QAOA --exclude=examples/IBM_QAOA/Archive --select=F401,F841` reports 15. Two worth a look beyond tidiness:

- `src/utils.py:1561` sets `y_scale = 100.0` and never uses it — in a plotting helper that computes `main_ymax` by hand just below, so it reads like an axis-scaling intent that was dropped.
- `src/Processing.py:19-23` imports `stochastic_benchmark as SB`, `bootstrap`, `interpolate`, `stats`, and `from utils_ws import *`, none used. The star-import in particular makes the module's namespace depend on the core package for no benefit.

The rest: `Processing.py:12` (`pickle`), `utils.py:11` (`Patch`), and unused locals at `utils.py:991,1049,1142,1154,1216,1909`.

### `tests/fixtures/ibm_qaoa/README.md` points at a path that does not exist 

Line 3 says the fixtures validate `examples/IBM_QAOA/ibm_qaoa_processing.py`. At this head that file is `examples/IBM_QAOA/Archive/ibm_qaoa_processing.py`, and the module the fixtures are actually exercised against is `examples/IBM_QAOA/src/Processing.py` (via `tests/test_ibm_qaoa_processing.py`). Please repoint it.

### PR description understates the change 

The body describes one notebook. The PR also adds a Nautilus/Kubernetes campaign harness, `run_prepare_pss_campaign.py`, a ~4.6k-line `simulation_validation.py`, a second notebook, ~600 KB of generated instance data, an `Archive/` tree, and two edits to core `src/training.py`. It links no issue and no commit carries a closing keyword. Please bring the body up to date with the actual scope and the verification you ran, and add a `Refs #...` if there is a tracking issue — the description is what reviewers and post-merge readers work from.

### `src/__init__.py` added with no explanation 

An empty `src/__init__.py` turns the `pythonpath = "src"` directory into a package as well. Nothing in the PR appears to need it, and it makes `import src.<mod>` resolvable from the repo root at the same time `tests/test_ibm_qaoa_processing.py` relies on `src` resolving to `examples/IBM_QAOA/src` — the test only works because it inserts `IBM_QAOA_ROOT` at `sys.path[0]`. If it is not needed, drop it; if it is, say what needs it.

### Pandas fixture stubs production objects with ad-hoc types 

`tests/test_stochastic_benchmark_pandas.py:32-33`

```python
sb.here = type('obj', (object,), {'checkpoints': str(tmp_path)})
sb.baseline = type('obj', (object,), {'recalibrate': lambda _df: None})()
```

This satisfies `get_experiment_parameters()` and the tests pass, so the blocker is cleared. But `here` is really a `names.paths(...)` and `baseline` a `VirtualBestBaseline`, and the stubs are inconsistent with each other (`here` is assigned the class, `baseline` an instance). Using `names.paths(tmp_path)` for `here` would exercise the real object shape at no extra cost.

## Questions

### External dependency installation is still undocumented 

The READMEs now name `QAOA-Parameter-Setting` and `qaoa_training_pipeline` and the three env vars, which answers most of the earlier question. Still missing is how to get the Python packages: `requirements-examples.txt` installs `scikit-learn`, `dimod`, `nbconvert`, `ipykernel` — no `qiskit`, no `qiskit-aer`. `nautilus/run_simulation_validation.sh:81-85` shows the real recipe (`pip install -e "${PIPELINE_REPO}" qiskit-aer`, then `pip install -e "${QPS_REPO}"`). Should that become an `examples/IBM_QAOA/requirements.txt`, or a setup block in `notebooks/README.md`?

### Should the generated instance data be committed? 

`examples/IBM_QAOA/data/generated_instances/heavy_hex__n=144__.../` adds 40 files (592 KB): 20 instance graphs at ~23 KB each plus 20 min/max-cut files. The code can regenerate these — `generate_training_instances_like_qps` and `_compute_and_write_minmax` — and the cache key is derived deterministically from graph type, node count, start index, and count. Is this committed deliberately as a reproducibility snapshot for the campaign, or is it scratch output that a user would regenerate locally? If it is deliberate, `notebooks/README.md:27` is the place to say so, since it currently describes the directory as a cache.

## Tests run and outcomes

Local `.venv` (Python 3.13.5, pandas 3.0.5, numpy 2.5.2, pytest 9.1.1), PR head `3bf76ec` in a clean worktree:

- `pytest tests/ -q` — **332 passed**, 12 warnings.
- `python run_tests.py unit` — **321 passed**; the command line includes the four zero-byte `*_groups.py` files.
- `flake8 src --select=E9,F63,F7,F82` — **0**.
- `flake8 examples/IBM_QAOA --select=E9,F63,F7,F82` — **23**, all in `Archive/ibm_qaoa_processing.py`.
- `flake8 examples/IBM_QAOA --exclude=Archive --select=E9,F63,F7,F82` — **0**; `--select=F401,F841` — **15**.
- `pytest examples/IBM_QAOA/Archive --collect-only` — collects 3 tests from `Archive/test_approx_ratio_fix.py`.
- Probe: restoring `include_groups=False` at `src/training.py:141,153` — **332 passed**.

Merge result (`origin/main` b35f31d + `3bf76ec`, temporary worktree): `git merge` clean, `pytest tests/ -q` — **447 passed**. `origin/main` alone: **426 passed**.

Baselines: pandas 2.3.3 and 3.0.5 probes for the `include_groups` finding, as described above.

GitHub checks on `3bf76ec`: `test (3.10)`, `test (3.11)`, `test (3.12)`, `integration-tests`, `coverage-summary`, `tutorial-smoke` — all pass. Note that none of them cover the two new notebooks, since they are absent from `examples/tutorials.json`.

Merge-readiness: the PR is mergeable and green, but it is 32 commits behind `main` and the diff (90 files, ~61k additions) exceeds GitHub's rendering limit, which is also why Copilot declined to review it. Splitting the `Archive/` tree, the committed data, and the core `src/training.py` edits into separate changes would make the remaining analysis workflow reviewable.

I would not merge this until the blocking issues above are addressed.

File: tests/test_stochastic_benchmark_pandas.py (hunk 1 above)
Comment (bernalde) [reply tests/test_stochastic_benchmark_pandas.py]: Addressed at `3bf76ec`. The fixture now populates the state and feeds `sb.get_experiment_parameters()` into `VirtualBestBaseline` / `ProjectionExperiment` instead of passing the `stochastic_benchmark` instance, and `checkpoint_path` resolves via `sb.here.checkpoints` under `tmp_path`. All 7 tests in this file pass (`pytest tests/ -q` — 332 passed at this head). Resolving.

One leftover nit, not worth reopening this thread: the fixture stubs `here` and `baseline` with anonymous `type(...)` objects rather than the real `names.paths(...)` / `VirtualBestBaseline`. I noted it as nonblocking in the review at this head.

File: examples/IBM_QAOA/src/Processing.py (hunk 2 above)
Comment (bernalde) [reply examples/IBM_QAOA/src/Processing.py:1]: Addressed at `3bf76ec`. `tests/test_ibm_qaoa_processing.py` now exists under the collected `tests/` path and covers the current modules rather than the archived `ibm_qaoa_processing` path: `set_data_path`, `load_problem_instance` degree filtering, `QAOATraining.locate_training_instance` / `load_training_instance` against the committed fixtures, `expected_energy_from_counts` (full/top modes plus the invalid-mode error), the min/max-cut and approximation-ratio helpers in `approx_ratio_calc`, and the PSS shard/filter/frontier helpers. 8 tests, all passing. Resolving.

File: examples/IBM_QAOA/src/simulation_validation.py (hunk 4 above)
Comment (bernalde) [reply examples/IBM_QAOA/src/simulation_validation.py]: The ask here is done at `3bf76ec`: the contributor-specific desktop default is gone, the defaults are derived from `__file__`, and `QAOA_PARAMETER_SETTING_ROOT` / `QAOA_TRAINING_PIPELINE_ROOT` / `IBM_QAOA_INSTANCE_CACHE_ROOT` are documented in `notebooks/README.md` and `src/README.md`. Resolving this thread.

To be clear about what is *not* settled: the replacement default resolves to the wrong directory. `simulation_validation.py:32-33` uses `parents[3]`, which is the repo root, so the defaults point at `<repo>/QAOA-Parameter-Setting` and `<repo>/qaoa_training_pipeline` — inside this checkout — while both READMEs document sibling checkouts next to it. That needs `parents[4]`, and I raised it as a separate blocking finding in the review at this head rather than keeping it on this thread.

**Comment by anurag-r20** [comment]
@bernalde

**Review by bernalde — APPROVED** [review]
Approving at `991c8e9`. This supersedes my `CHANGES_REQUESTED` review at `3f2d783` — every finding from that round is now verified fixed at this head.

**Disclosure:** the final commit `991c8e9` is mine (two lines, replacing the contributor-path fallbacks in `Analysis.ipynb`), so I am approving a PR whose last commit I wrote. It is a two-line change in an 82-file PR and the rest is Anurag's work, but it should be on the record.

## Findings from the last round, at this head

| Finding | Status |
|---|---|
| `nautilus-personal-fork-url` (**Blocking**) | Fixed — all 8 manifests use `usra-riacs/stochastic-benchmark.git`; zero fork references tree-wide |
| `contributor-paths-in-committed-artifacts` | Fixed — `/mnt/c/Users/…` now returns **0 hits anywhere in the tree**. `4397260` scrubbed the 20 min/max-cut data files and changed the producer to `Path(graph_path).name`; `d912fb7` stopped `save_current_plot` printing an absolute path; `991c8e9` replaced the two `Analysis.ipynb` fallbacks with the documented sibling layout |
| `pr-description-understates-scope` | Fixed — body rewritten to the real scope, and every claim in its test plan reproduces (see below) |
| `root-src-init-py-unexplained` | Fixed — `src/__init__.py` removed |
| `pandas-fixture-stubs-adhoc-types` | Fixed — `sb.here = names.paths(str(tmp_path))` |
| `pr-84-stacking-coordination` | Fixed — #84 retargeted to `QAOA_Parameter_Setting_IBM` |
| `example-dependency-setup` (Question) | Answered — `qiskit-aer==0.17.2` in `requirements-examples.txt`, plus a setup section in `notebooks/README.md` |
| `committed-generated-instance-data` (Question) | Answered — `notebooks/README.md` now states the snapshot is deliberate |

Two commits since that round were unprompted fixes rather than review responses, and both check out:

- `a7cd6f1` — `_relativize_paths` raised `ValueError` for any spec row whose `graph_path`/`minmax_path` pointed into the sibling `QAOA-Parameter-Setting` checkout rather than under `REPO_ROOT`. The fix rebases on `WORKSPACE_ROOT` (defined as `REPO_ROOT.parent` in the same notebook) and falls back to leaving a path unscrubbed instead of crashing. Correct and appropriately defensive for what is a display convenience.
- `d912fb7` — `save_current_plot` now prints a cwd-relative path when given an absolute `plot_dir`.

## Tests run and outcomes

**At PR head `991c8e9`** (clean worktree, Python 3.13.5 / pandas 3.0.5 / numpy 2.5.2):

- `pytest tests/ -q` — **333 passed**
- `python run_tests.py unit` — passed
- `flake8 src --select=E9,F63,F7,F82` — **0**
- `flake8 --select=E9,F63,F7,F82,F401,F841 examples/IBM_QAOA` — **0**
- `scripts/verify_tutorials.py --dry-run` — all 9 notebooks resolve; both IBM_QAOA notebooks correctly `SKIP … (external)` with reasons
- Both notebooks: 0 error outputs, 0 unexecuted cells, monotonic execution counts, **0 absolute paths in any output**
- Changed-file set matches GitHub's exactly (82/82)

Every line of the PR body's test plan reproduces.

**Merge result** (`origin/main` `b35f31d` + `991c8e9`, temporary worktree): merge is clean, `pytest tests/ -q` — **448 passed**. `origin/main` alone — **426 passed**.

**GitHub checks on `991c8e9`:** `test (3.10)`, `test (3.11)`, `test (3.12)`, `tutorial-smoke`, `integration-tests`, `coverage-summary` — 6/6 pass, none skipped.

## Merge readiness

No blocking issues, no nonblocking issues, no open questions — 0 findings in this review, and 0 unresolved review threads.

The branch is 32 commits behind `main`, which is why the merge-result run above matters more than the head run; both are green.

One coordination note for after the merge: #84 targets `QAOA_Parameter_Setting_IBM`, so leaving that branch in place keeps #84 stacked and collapses its diff to its own 15 files. #84 is not ready independently — it still needs a base merge to clear a notebook conflict and a `pytest.importorskip("qiskit_aer")` guard on `TestBuildBoundCircuitSimulator`, whose four tests fail in any environment without qiskit installed.
